sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala

```diff
@@ -76,28 +76,34 @@ abstract class PartitioningAwareFileIndex(
     // be applied to files.
     val fileMetadataFilterOpt = dataFilters.filter { f =>
       f.references.nonEmpty && f.references.forall {
-        case FileSourceMetadataAttribute(_) => true
+        case FileSourceConstantMetadataAttribute(_) => true
```

Comment:
After this change, now we exclude any unflattened metadata predicates. This is probably why the new test passes

Reply:
Here we do expect to only see FileSourceContant/GeneratedMetadata attributes. It is hard to test this though in an automated fashion, and I think actually adding an assert here is not the best idea (not filtering is a better option than throwing an exception).
I manually verified this for now